library/core/src/ptr/const_ptr.rs

```diff
@@ -101,6 +99,70 @@ impl<T: ?Sized> *const T {
         if self.is_null() { None } else { unsafe { Some(&*self) } }
     }
 
+    /// Returns `None` if the pointer is null, or else returns a shared reference to
+    /// the value wrapped in `Some`.
+    ///
+    /// # Safety
+    ///
+    /// When calling this method, you have to ensure that *either* the pointer is NULL *or*
+    /// all of the following is true:
+    ///
+    /// * The pointer must be properly aligned,
+    ///
+    /// * it must be "dereferencable" in the sense defined in [the module documentation], and
+    ///
+    /// * the lifetime `'a` returned is arbitrarily chosen and does
+    ///   not necessarily reflect the actual lifetime of the data. *You* must enforce
+    ///   Rust's aliasing rules. In particular, for the duration of this lifetime,
+    ///   the memory the pointer points to must not get mutated (except inside `UnsafeCell`).
+    ///
+    /// This applies even if the result of this method is unused!
+    ///
+    /// [the module documentation]: crate::ptr#safety
+    ///
+    /// # Examples
+    ///
+    /// Basic usage:
+    ///
+    /// ```
+    /// #![feature(ptr_as_uninit)]
+    ///
+    /// let ptr: *const u8 = &10u8 as *const u8;
+    ///
+    /// unsafe {
+    ///     if let Some(val_back) = ptr.as_uninit_ref() {
+    ///         println!("We got back the value: {}!", val_back.assume_init());
+    ///     }
+    /// }
+    /// ```
+    ///
+    /// # Null-unchecked version
+    ///
+    /// If you are sure the pointer can never be null and are looking for some kind of
+    /// `as_uninit_ref_unchecked` that returns the `&MaybeUninit<T>` instead of
+    /// `Option<&MaybeUninit<T>>`, know that you can dereference the pointer directly.
+    ///
+    /// ```
+    /// use std::mem::MaybeUninit;
+    ///
+    /// let ptr: *const u8 = &10u8 as *const u8;
+    ///
+    /// unsafe {
+    ///     let val_back = &*(ptr as *const MaybeUninit<u8>);
```

Comment:
I am always hesitant to recommend pointer casts. They are a very dangerous operation as the type checker checks very little about them.

There's no unwrap_unchecked, is there?

TBH I never understood why NULL was treated separately from the other erorr conditions in these methods, but I expect it is too confusing to make the uninit versions return a plain reference without `Option`?

Reply:
I just copied the previous example but I don't like the example either. Actually the method is unsafe anyway, so dereferencing a raw pointer in this context seems ok'ish.

No, sadly, there is no `unwrap_unchecked` but there should be IMO. If we add it, this example should definitely be changed.

Yes, that was the reason about it. I'd expect `as_uninit_ref` to behave exactly as `as_ref`, besides wrapped in an `MaybeUninit`. Also it is UB to have `&MaybeUninit<T>` pointing to NULL. For a pointer guaranteed to be non-NULL we have `NonNull`, which should be used whenever possible IMO.

Reply:
The example you copied doesn't need a pointer cast though.

Reply:
Oh, I'm sorry. Maybe we should just leave that out then?

Reply:
It's still a fair point though, there should be an ergonomic way to do that, and IMO "ergonomic" implies "no `as`".

I am fine with punting on that for now and leaving a note in the tracking issue.